log/DESIGN.md

```diff
@@ -0,0 +1,671 @@
+
+The required `name` parameter is accepted as a `string` method argument.
+
+The following options are defined to support optional parameters:
+
+```go
+func WithInstrumentationVersion(version string) LoggerOption
+func WithInstrumentationAttributes(attr ...attribute.KeyValue) LoggerOption
+func WithSchemaURL(schemaURL string) LoggerOption
+```
+
+Implementation requirements:
+
+- The [specification requires](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#concurrency-requirements)
+  the method to be safe to be called concurrently.
+
+- The method should use some default name if the passed name is empty
+  in order to meet the [specification's SDK requirement](https://opentelemetry.io/docs/specs/otel/logs/sdk/#logger-creation)
+  to return a working logger when an invalid name is passed
+  as well as to resemble the behavior of getting tracers and meters.
+
+`Logger` can be extended by adding new `LoggerOption` options
+and adding new exported fields to the `LoggerConfig` struct.
+This design is already used in Trace API for getting tracers
+and in Metrics API for getting meters.
+
+Rejected alternative:
+
+- [Passing struct as parameter to LoggerProvider.Logger](#passing-struct-as-parameter-to-loggerproviderlogger).
+
+### Logger
+
+The [`Logger` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#logger)
+is defined as an interface:
+
+```go
+type Logger interface {
+	embedded.Logger
+	Emit(ctx context.Context, record Record)
+}
+```
+
+The specification may add new operations to `Logger`.
+The interface may have methods added without a package major version bump.
+This embeds `embedded.Logger` to help inform an API implementation
+author about this non-standard API evolution.
+This approach is already used in Trace API and Metrics API.
+
+### Logger.Emit
+
+The `Emit` method implements the [`Emit a LogRecord` operation](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord).
+
+[`Context` associated with the `LogRecord`](https://opentelemetry.io/docs/specs/otel/context/)
+is accepted as a `context.Context` method argument.
+
+Calls to `Emit` are supposed to be on the hot path.
+Therefore, in order to reduce the number of heap allocations,
+the [`LogRecord` abstraction](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#emit-a-logrecord),
+is defined as a `Record` type:
+
+```go
+type Record struct {
+	timestamp         time.Time
+	observedTimestamp time.Time
+	severity          Severity
+	severityText      string
+	body              Value
+
+	// The fields below are for optimizing the implementation of
+	// attributes.
+	front [5]KeyValue
+	nFront int // The number of attributes in front.
+	back []KeyValue
+}
+```
+
+[`Timestamp`](https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-timestamp)
+is accessed using following methods:
+
+```go
+func (r Record) Timestamp() time.Time 
```

Comment:
From https://go.dev/tour/methods/8:

> In general, all methods on a given type should have either value or pointer receivers, but not a mixture of both. (We'll see why over the next few pages.)

From https://go.dev/wiki/CodeReviewComments#receiver-type:

> Don’t mix receiver types. Choose either pointers or struct types for all available methods.

However, [`slog.Record`](https://pkg.go.dev/log/slog#Record) mixes the receiver types to improve performance. See: https://github.com/golang/go/issues/56345#issuecomment-1302563756

>> some receiver of Record struct is by value

> Passing Records by value means they incur no heap allocation. That improves performance overall, even though they are copied.

Also from https://go.dev/wiki/CodeReviewComments#receiver-type:

> there are times when a value receiver makes sense, usually for reasons of efficiency, such as for small unchanging structs

I think it should be documented. I can try doing some benchmarks, but I think it may be hard as I guess compiler optimization can kick in.

Let me know what you think and if you have any concerns.

Reply:
SIG meeting:
We should have a proof for https://github.com/golang/go/issues/56345#issuecomment-1302563756.
I will work on it.

Reply:
I created https://github.com/pellared/opentelemetry-go/pull/8 and the benchmarks run on Go 1.21 do not show any noticeable differences.

@jba, are you able help and share what is the scenario when having all methods of `slog.Record` with a pointer receiver would cause a heap allocation? Thanks in advance.

Reply:
Addressed https://github.com/open-telemetry/opentelemetry-go/pull/4809/commits/499da4e003d13522f02ba6813aaabfc30691d5a7

Keeping the conversation open for a few days to give a chance for @jba to comment.

Reply:
Hi, sorry for the delay, I was on vacation.

The guideline to avoid mixing receiver types is a good one, but it is just that—a guideline. There are exceptions.

Record is a value type, so that calls to `slog.Handler.Handle` do not result in heap allocation. So all of `Record`s methods should have value receivers. But some methods change the receiver, so they have to have pointer receivers.

To answer your question, I'm not sure if making all the methods have pointer receivers would affect allocation or performance. Probably not significantly. But `Record` has a mix for the reason I gave above.